src/utils.c

```diff
@@ -313,12 +344,43 @@ ts_timestamptz_bucket(PG_FUNCTION_ARGS)
 {
 	Interval   *interval = PG_GETARG_INTERVAL_P(0);
 	TimestampTz timestamp = PG_GETARG_TIMESTAMPTZ(1);
+	TimestampTz origin;
 	TimestampTz result;
 	int64		period = -1;
 
+
+	if (PG_NARGS() > 2)
+		origin = PG_GETARG_TIMESTAMPTZ(2);
+	else
+	{
+		/*
+		 * The default origin moves the PG epoch to start on a monday:
+		 * 2000-01-03. We don't use PG epoch since it starts on a saturday.
+		 * This makes time-buckets by a week more intuitive and aligns it with
+		 * date_trunc.
+		 */
+#ifdef HAVE_INT64_TIMESTAMP
+		origin = 2 * USECS_PER_DAY;
+#else
+		origin = 2 * SECS_PER_DAY;
+#endif
+	}
+
 	if (TIMESTAMP_NOT_FINITE(timestamp))
 		PG_RETURN_TIMESTAMPTZ(timestamp);
 
+	if (origin > 0 && timestamp < DT_NOBEGIN + origin)
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				 errmsg("timestamp out of range")));
+
+	if (origin < 0 && timestamp > DT_NOEND + origin)
+		ereport(ERROR,
+				(errcode(ERRCODE_NUMERIC_VALUE_OUT_OF_RANGE),
+				 errmsg("timestamp out of range")));
+
+	timestamp -= origin;
+
 	period = get_interval_period_timestamp_units(interval);
```

Comment:
I think we should error on negative interval as that breaks alot of assumptions in other places.

Reply:
Fixed